Add support for parsing Web Extension commands.

xeenon opened the pull request:
#### 210a20b544ef3f77cea2f7dcb19ec21b9022d5cf
<pre>
Add support for parsing Web Extension commands.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260157">https://bugs.webkit.org/show_bug.cgi?id=260157</a>
<a href="https://rdar.apple.com/problem/114823250">rdar://problem/114823250</a>

Reviewed by Brian Weinstein.

Add support for parsing Web Extension commands and the new `_WKWebExtensionCommand`.
This does not expose the `browser.commands` API yet, that will be next.

The `_WKWebExtensionCommand` class supports customization by the app, which will
be reflected back to the extensions in `browser.commands.getAll()`.

* Source/WebCore/en.lproj/Localizable.strings: Updated.
* Source/WebKit/Modules/OSX_Private.modulemap: Added new _WKWebExtensionCommand.
* Source/WebKit/Modules/iOS_Private.modulemap: Ditto.
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension hasCommands]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommand.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommand.mm: Added.
(-[_WKWebExtensionCommand dealloc]):
(-[_WKWebExtensionCommand hash]):
(-[_WKWebExtensionCommand isEqual:]):
(-[_WKWebExtensionCommand description]):
(-[_WKWebExtensionCommand debugDescription]):
(-[_WKWebExtensionCommand webExtensionContext]):
(-[_WKWebExtensionCommand identifier]):
(-[_WKWebExtensionCommand discoverabilityTitle]):
(-[_WKWebExtensionCommand activationKey]):
(-[_WKWebExtensionCommand setActivationKey:]):
(-[_WKWebExtensionCommand modifierFlags]):
(-[_WKWebExtensionCommand setModifierFlags:]):
(-[_WKWebExtensionCommand _apiObject]):
(-[_WKWebExtensionCommand _webExtensionCommand]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommandInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommandPrivate.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext commands]):
(-[_WKWebExtensionContext performCommand:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::toAPI):
(WebKit::WebExtension::createError):
(WebKit::WebExtension::errors):
(WebKit::WebExtension::commands):
(WebKit::WebExtension::hasCommands):
(WebKit::parseCommandShortcut):
(WebKit::WebExtension::populateCommandsIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm: Added.
(WebKit::WebExtensionCommand::setActivationKey):
(WebKit::WebExtensionCommand::setModifierFlags):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::commands):
(WebKit::WebExtensionContext::performCommand):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.cpp: Added.
(WebKit::WebExtensionCommand::WebExtensionCommand):
(WebKit::WebExtensionCommand::operator== const):
(WebKit::WebExtensionCommand::extensionContext const):
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.h: Added.
(WebKit::WebExtensionCommand::create):
(WebKit::WebExtensionCommand::identifier const):
(WebKit::WebExtensionCommand::description const):
(WebKit::WebExtensionCommand::activationKey const):
(WebKit::WebExtensionCommand::modifierFlags const):
(WebKit::WebExtensionCommand::wrapper const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionContext.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/270494@main">https://commits.webkit.org/270494@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b488c5e477e0892d9f4b2d2083a715f8bbcbeb95

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25616 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4221 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26899 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27714 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25899 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5968 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1656 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23607 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25865 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3138 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22079 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28296 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2775 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23029 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29125 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23375 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23395 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26971 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2795 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1028 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4162 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3235 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3273 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3110 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommand.h:
```diff
@@ -0,0 +1,85 @@
+/*!
+ @abstract A `WKWebExtensionCommand` object encapsulates the properties for an individual web extension command.
+ @discussion Provides access to command properties such as a unique identifier, a descriptive title, and shortcut keys. Commands
+ can be used by a web extension to perform specific actions within a web extension context, such toggling features, or interacting with
```
Super nit: The new sentence starting with Commands could start on this line.

b-weinstein replied:
Super nit: The new sentence starting with Commands could start on this line.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommand.mm:
```diff
@@ -0,0 +1,174 @@
+
+- (NSString *)description
+{
+    return self.discoverabilityTitle;
```
Should the description have the extension display name/identifier as well?

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    if (!stringBuilder.isEmpty())
+        stringBuilder.append('+');
```
Do we want this to fail if no modifiers were specified?

xeenon committed:
Add support for parsing Web Extension commands.
https://bugs.webkit.org/show_bug.cgi?id=260157
rdar://problem/114823250

Reviewed by Brian Weinstein.

Add support for parsing Web Extension commands and the new `_WKWebExtensionCommand`.
This does not expose the `browser.commands` API yet, that will be next.

The `_WKWebExtensionCommand` class supports customization by the app, which will
be reflected back to the extensions in `browser.commands.getAll()`.

* Source/WebCore/en.lproj/Localizable.strings: Updated.
* Source/WebKit/Modules/OSX_Private.modulemap: Added new _WKWebExtensionCommand.
* Source/WebKit/Modules/iOS_Private.modulemap: Ditto.
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension hasCommands]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommand.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommand.mm: Added.
(-[_WKWebExtensionCommand dealloc]):
(-[_WKWebExtensionCommand hash]):
(-[_WKWebExtensionCommand isEqual:]):
(-[_WKWebExtensionCommand description]):
(-[_WKWebExtensionCommand debugDescription]):
(-[_WKWebExtensionCommand webExtensionContext]):
(-[_WKWebExtensionCommand identifier]):
(-[_WKWebExtensionCommand discoverabilityTitle]):
(-[_WKWebExtensionCommand activationKey]):
(-[_WKWebExtensionCommand setActivationKey:]):
(-[_WKWebExtensionCommand modifierFlags]):
(-[_WKWebExtensionCommand setModifierFlags:]):
(-[_WKWebExtensionCommand _apiObject]):
(-[_WKWebExtensionCommand _webExtensionCommand]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommandInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionCommandPrivate.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext commands]):
(-[_WKWebExtensionContext performCommand:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::toAPI):
(WebKit::WebExtension::createError):
(WebKit::WebExtension::errors):
(WebKit::WebExtension::commands):
(WebKit::WebExtension::hasCommands):
(WebKit::parseCommandShortcut):
(WebKit::WebExtension::populateCommandsIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm: Added.
(WebKit::WebExtensionCommand::setActivationKey):
(WebKit::WebExtensionCommand::setModifierFlags):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::commands):
(WebKit::WebExtensionContext::performCommand):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.cpp: Added.
(WebKit::WebExtensionCommand::WebExtensionCommand):
(WebKit::WebExtensionCommand::operator== const):
(WebKit::WebExtensionCommand::extensionContext const):
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.h: Added.
(WebKit::WebExtensionCommand::create):
(WebKit::WebExtensionCommand::identifier const):
(WebKit::WebExtensionCommand::description const):
(WebKit::WebExtensionCommand::activationKey const):
(WebKit::WebExtensionCommand::modifierFlags const):
(WebKit::WebExtensionCommand::wrapper const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionContext.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/270494@main
